File: Java/src/RecurssionBacktracking/Permutations.java
package RecurssionBacktracking;

import java.util.ArrayList;

// Time complexity: O(n!n)
// Space complexity: O(n)

public class Permutations {
    public static void main(String[] args) {
        int[] arr = { 1, 2, 3 };
        printPermutation(arr, 0, new boolean[arr.length], new ArrayList<>());
    }

    public static void printPermutation(int[] arr, int pos, boolean[] selected, ArrayList<Integer> ans) {
        if (pos == arr.length) {
            System.out.println(ans);
            return;
        }
        for (int i = 0; i < arr.length; i++) {
            if (selected[i] == false) {
                selected[i] = true;
                ans.add(arr[i]);
                printPermutation(arr, pos + 1, selected, ans);
                selected[i] = false;
                ans.remove(ans.size() - 1);
            }
        }
    }
}

File: Python/RecurssionBacktracking/Permutations.py

# Time complexity: O(n!n)
# Space complexity: O(n)

def main():
    arr = [1, 2, 3]
    selected = [False for i in range(len(arr))]
    ans = []
    printPermutations(arr, 0, selected, ans)


def printPermutations(arr, pos, selected, ans):
    if (pos == len(arr)):
        print(ans)
        return
    for i in range(len(arr)):
        if (selected[i] == False):
            selected[i] = True
            ans.append(arr[i])
            printPermutations(arr, pos+1, selected, ans)
            selected[i] = False
            ans.pop()


Solution()
